src/param.rs

```diff
@@ -0,0 +1,92 @@
+use failure::Error;
+use libnest::package::{CategoryName, PackageName};
+use rocket::http::RawStr;
+use rocket::request::FromParam;
+use semver::Version;
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct CategoryNameParam {
+    value: CategoryName,
+}
+
+impl CategoryNameParam {
+    pub fn value(&self) -> &CategoryName {
+        &self.value
+    }
+}
+
+impl<'a> FromParam<'a> for CategoryNameParam {
+    type Error = Error;
+
+    fn from_param(param: &'a RawStr) -> Result<Self, Self::Error> {
+        let decoded_param = param.percent_decode()?;
+
+        Ok(CategoryNameParam {
+            value: CategoryName::parse(decoded_param.as_ref())?,
+        })
+    }
+}
+
+impl Into<CategoryName> for CategoryNameParam {
+    fn into(self) -> CategoryName {
+        self.value
+    }
+}
+
+#[derive(Clone, Eq, PartialEq, Ord, PartialOrd, Hash, Debug)]
+pub struct PackageNameParam {
+    value: PackageName,
+}
+
+impl PackageNameParam {
```

Comment:
I'm not a fan of macro to replace lazyness :D

It's not a big deal here, not like a thousand of lines, only a few dozen, so I'd like to keep it this way.

Reply:
That's where I don't agree. I don't suggest adding macros out of laziness, but as a way to **factor common behavior rather than repeating it**.

Since in this case, we cannot use a trait for that, we fall back on macros, which are a totally appropriate tool. If you feel particularly adventurous, you could even go with a cool procedural macro, and get some **good-guy :tm: karma points** :D